File: web/src/types/Despesa.ts
import { getDataAtual } from "../utils/DateUtils";
import { Categoria } from "./Categoria";
import { Parcela } from "./Parcela";

export type Despesa = {
  id: number;
  categoria: Categoria;
  dataLancamento: Date;
  descricao: string;
  valorTotal: number;
  situacao: string;
  parcelas: Parcela[];
}

export const initialDespesaState: Despesa = {
  id: 0,
  categoria: { id: 0, descricao: '' },
  dataLancamento: getDataAtual(),
  descricao: '',
  valorTotal: 0,
  situacao: '',
  parcelas: []
};

File: server/src/main/java/br/com/lcano/centraldecontrole/domain/Despesa.java
package br.com.lcano.centraldecontrole.domain;

import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import jakarta.persistence.*;
import lombok.*;
import java.io.Serial;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Table(name = "despesa")
@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(of = "id")
@ToString(of = "id")
public class Despesa implements Serializable {
    @Serial
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "idusuario", nullable = false)
    private Usuario usuario;

    @ManyToOne
    @JoinColumn(name = "idcategoria", nullable = false)
    private CategoriaDespesa categoria;

    @Column(nullable = false)
    private String descricao;

    @Column(name = "datalancamento", nullable = false)
    private Date dataLancamento;

    @OneToMany(mappedBy = "despesa", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<DespesaParcela> parcelas = new ArrayList<>();
}
